### ISSUE TITLE
[UX] Permissions matrix cut off on Permissions tab of Create Content page

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#5384]
## Description of the bug

This was [first reported in the Field Permissions issue queue](https://github.com/backdrop-contrib/field_permissions/issues/6), but it's actually a core issue.

If there are many roles, then on the Permissions tab of the Add Content Type page (admin/structure/types/add), some of the permissions are cut off on the right side with no way to reach them.

## Steps To Reproduce

To reproduce the behavior:
1. Create one or more roles with really long single-word names, e.g., "Supercalifragilisticexpialodocious."
2. Go to admin/structure/types/add.
3. Click on the Permissions tab.

## Actual behavior

Rightmost permissions are cut off.

![sample](https://user-images.githubusercontent.com/29214084/144732419-0f042841-6b88-4fcf-b3d6-597b759df22c.jpg)

## Expected behavior

The permissions matrix should scroll as needed, so a little CSS tweak should fix it.

**Comment by indigoxela** [comment]
Great, this now works properly! On smaller screens the default roles have been cut off already, now everything is accessible.

**Comment by indigoxela** [comment]
One small glitch appears with the sticky header - but I'm not sure if that can get fixed. The scrolling table and the sticky header table seem sort of disconnected. Sticky table headers aren't very mobile-friendly, anyway (maybe a follow-up?).

![sticky-header-problem](https://user-images.githubusercontent.com/842066/144985063-be447875-e335-4aeb-9395-1863046b039d.png)

**Comment by klonos** [comment]
The content type edit form is not the only place where this issue can happen - we have such permission matrices in places like `admin/structure/taxonomy/%/configure`. So I think that whatever change we get in place should cover all cases (and most likely be in system module instead of node).

**Comment by bugfolder** [comment]
Makes sense.

It looks like things in core that generate a permissions matrix reuse `theme_user_admin_permissions()` (in `user.theme.inc`) to create the matrix. That includes the callbacks for `contact_form_permissions`, `node_type_form_permissions`, `taxonomy_vocabulary_form_permissions`, and of course, the user permissions page itself. So that would suggest since `user` module is generating the table, the changes should happen there. Reasonable?

**Comment by bugfolder** [comment]
Revised PR submitted. This new PR only changes CSS in the `user` module, and addresses all the places I could find where the permissions matrix appears:

* admin/config/people/permissions
* admin/structure/types/add and admin/structure/types/manage/%
* admin/structure/taxonomy/add and admin/structure/taxonomy/%/configure
* admin/structure/contact/settings

As @indigoxela noted, there is an issue with sticky headers, but that's a broader issue that affects pages even outside of this PR (for example, admin/config/people/permissions). So I'll submit a follow-up issue.

**Comment by indigoxela** [comment]
@bugfolder many thanks for this alternate PR. It works for me ... for permission matrices.

![side-scrolling-taxo](https://user-images.githubusercontent.com/842066/146143546-0ff1a250-28db-4a19-8496-c636559c968d.png)

There are more situations where tables on smaller screens get cut off, especially but not only in vertical tabs. However, following the issue topic, we're only trying to fix permission matrices.

**Comment by docwilmot** [comment]
This also works for me, but I am no CSS expert so cant mark RTBC. But I suspect it is. Added milestone though.

**Comment by indigoxela** [comment]
I did final testing with node type and vocab forms, contact settings and the main permissions matrix. That all looks good. For the sticky header, which will be a bigger task, we already have a follow-up.

The added CSS is minimal and specific - it only applies to few form items. And I don't think that an overflow definition like that can ever do harm. (CSS changes can be tricky, that's why I'm cautious.)

RTBC :+1:

**Comment by klonos** [comment]
Curious to see what our Drupal brethren will come up with in the related (not exactly the same though) https://www.drupal.org/project/drupal/issues/2152525.

Thanks @bugfolder 🙏🏼  ...this works much better now, but I would like us to ideally see if we can make the first column (permission names) sticky as well. Something like https://css-tricks.com/a-table-with-both-a-sticky-header-and-a-sticky-first-column. What does everyone else think? Follow-up, or work on it here?

It seems that there is wide browser support for `position: sticky` (everything besides IE really): https://caniuse.com/css-sticky

**Comment by bugfolder** [comment]
> Follow-up, or work on it here?

Since the misalignment happens in many places other than here, I'd already submitted a follow-up issue https://github.com/backdrop/backdrop-issues/issues/5396.

**Comment by klonos** [comment]
Thanks @bugfolder 🙏🏼 

I just noticed that although this works fine in the vertical tabs (only the permissions table becomes left/right scrollable), in `admin/config/people/permissions` the entire form (including the form submit button) becomes scrollable. Do you think you could perhaps tweak things so that only the permissions table has horizontal scrollbars?

**Comment by bugfolder** [comment]
> Do you think you could perhaps tweak things so that only the permissions table has horizontal scrollbars?

I don't see a way here. The problem is that the sticky header is set to a large width (presumably by the sticky-header processing), and that makes the enclosing form have a large width, so that the entire form scrolls in narrow windows. This prevents the table (inside the form) scrolling from kicking in.

The entire permissions sub-form scrolls in the other places this permissions table shows up, but since the submit button is outside the sub-form in those places, you don't see the submission button scrolling out of view.

So fixing that means fixing the (pre-existing) issues with the sticky header, which I think is a job for https://github.com/backdrop/backdrop-issues/issues/5396.

**Comment by klonos** [comment]
> I don't see a way here.

Perhaps I didn't do a good job explaining. What I meant was instead of this:
![issue5384-pr3863](https://user-images.githubusercontent.com/2423362/147505520-029910e9-7250-4c8f-8a51-ff3fc184675f.gif)
Have it behave like this:
![issue5384-only-sroll-permissions](https://user-images.githubusercontent.com/2423362/147505532-9626ab22-bac1-4c7a-9358-1c504114b3c9.gif)

If you agree, then I've filed a PR against your branch here: https://github.com/bugfolder/backdrop/pull/2 (the actual commit is [bugfolder/backdrop@`4a7928f` (#2)](https://github.com/bugfolder/backdrop/pull/2/commits/4a7928f9a266d0f166ec34f52c8ba2f202c917c9) - the rest are recent commits to core from other issues)

**Comment by bugfolder** [comment]
Agreed, nice. Made the changes directly (to avoid dealing with all the other changes in the PR against the branch).

**Comment by klonos** [comment]
Thanks @bugfolder 🙏🏼  ...things work much better now 👍🏼 

I would still want someone to take a look at the new class introduced, and the specificity (or not) of the css we're introducing, since I'm no expert.

PS: perhaps add an inline comment in the wrapping div we're introducing around the table, to explain why it's there (otherwise it may seem like an arbitrary decision). That way, people will know not to try to remove it in the future (or if they do, that they need to make sure that this issue does not regress).

**Comment by bugfolder** [comment]
When I first did the PR, I made it CSS-only to minimize possible side effects that could arise from changing the HTML structure. But now that we're adding the extra div wrapper and class, we can simplify the CSS to apply just to that class, `permissions-matrix`, and that will work for all current usages.

`permissions-matrix` isn't used anywhere else in core, so it looks safe to use as a CSS class. One possible question is whether `permissions-wrapper` would be a better name for the wrapper class. (That, too, is not yet in use anywhere in core.)

There is the question of whether adding an extra layer of `div` is going to break some theme somewhere, which seems like a valid question, but I do note that we rearrange the HTML of admin pages not-infrequently, so there is a precedent. But of course, I, too, welcome input from those with more experience in this area.

At any rate, I've added an inline comment, and took out the CSS that's no longer needed with the new wrapper.

Pages affected include:

* admin/config/people/permissions
* admin/structure/types/manage/*/configure
* admin/structure/contact/settings
* admin/structure/taxonomy/*/configure
